File: hackerrank/problem solving/algorithms/2 - implementation/11 - sales by match/with morfo library/sales_by_match.rb
# https://www.hackerrank.com/challenges/sock-merchant/problem?isFullScreen=true

require 'morfo'

def main
  Morfo.skip_input_ln
  socks_to_counts = Morfo.readln(&:to_i).tally
  puts sock_merchant(socks_to_counts)
end

# k: length of entries in socks_to_counts
# T: O(k)
# S: O(1) extra space
def sock_merchant(socks_to_counts) = socks_to_counts.each_value.sum { _1 / 2 }

main if __FILE__ == $0

File: hackerrank/problem solving/algorithms/2 - implementation/11 - sales by match/with morfo library/sales_by_match.java
// https://www.hackerrank.com/challenges/sock-merchant/problem?isFullScreen=true
// Java 25

import morfo.collections.Counter;
import morfo.io.Reader;

void main() {
    Reader.skipInputLn();
    var socksToCounts = counter(Reader.readLnInts());
    IO.println(sockMerchant(socksToCounts));
}

Map<Integer, Integer> counter(int[] arr) {
    return Counter.of(Arrays.stream(arr).boxed().toList());
}

// k: length of entries in socksToCounts
// T: O(k)
// S: O(1) extra space
int sockMerchant(Map<Integer, Integer> socksToCounts) {
    return socksToCounts.values().stream().mapToInt(x -> x / 2).sum();
}
